# Lab 02 - Sifting Weather Data

## Prelab

Complete Example 3.3 (p. 89) from the textbook.  Coding this example will help you with the lab assignment.  You can find the relevant .xls-file posted along with these lab instructions on D2L.  When importing the .xls-file, select the "Numeric Matrix" option.  After completing Example 3.3, save your script for this example to an M-file and set it aside.  Your work completing Example 3.3 does not need to be submitted for credit.

## Lab

Clear your variables (\texttt{clc, clear}) and re-import the weather data from the \textbf{weather\_data.xls} file into a variable called \textbf{wd}.  This time, keep the values of -99999 in the file.  Transpose the data (as done in Example 3.3) so the data for each month is in \textit{columns}.  Save your \textbf{wd} variable to a file called \textbf{weather\_data} using the following command.

```
import pandas as pd
df = pd.read_xls('weather_data.xlsx')
```

Create a new Jupuyter notebook called lab2. \input{.boiler_plate/include_file_header.tex} Below the header, start a new section with \%\%, and create a matrix \textbf{wd} from your \textbf{weather\_data.xls} file.  

You should create a pandas dataframe matrix \textbf{wd} with the ```pd.read_excel()``` function. Search the Pandas pacakge documentation help section to see how to use the ```pd.read_excel()``` function.  
Note: the .xlsx-file must be in your current file path, and single quotes need to surround the file name.  

In [13]:
import pandas as pd
df = pd.read_excel('weather_data.xlsx', header=None)
df.shape


(12, 31)

Afterwards, a running ```df.shape`` should indicate that the \textbf{wd} matrix has dimensions of 11 x 31.  Also check the value in row 1, column 3 in the dataframe is -99999 by running the ```df.head()``` method. 

In [14]:
df.head()

0    1    2   3   4    5   6   7   8   9   ...    21  22   23  24  25  26  \
0    0    0  272   0   0    0   0   9  26   1  ...     1  15  224   0   0   0   
1   61  103    0   2   0    0   0   2   0  68  ...     0   0    5   0   0   0   
2    2    0   17  27   0   10   0   0  30  25  ...    26   0    0   9   6  78   
3  260    1    0   0   0    1   3   0   5   0  ...     0   0    0   0   9  35   
4   47    0    0   0   5  115  49  81   0   0  ...     0   0    0   0   0   2   

   27     28     29     30  
0   0      0     33     33  
1  62 -99999 -99999 -99999  
2   0      5      8      0  
3  13     86      0 -99999  
4   0      0      0      0  

[5 rows x 31 columns]

Use the following line to change all of the values in the matrix \textbf{wd} which are less than zero (the -99999 values) and set them to zero:

```
df[df==-99999]=0
```

In [15]:
import numpy as np
df[df==-99999]=np.nan
df.head()

0    1    2   3   4    5   6   7   8   9   ...   21  22   23  24  25  26  \
0    0    0  272   0   0    0   0   9  26   1  ...    1  15  224   0   0   0   
1   61  103    0   2   0    0   0   2   0  68  ...    0   0    5   0   0   0   
2    2    0   17  27   0   10   0   0  30  25  ...   26   0    0   9   6  78   
3  260    1    0   0   0    1   3   0   5   0  ...    0   0    0   0   9  35   
4   47    0    0   0   5  115  49  81   0   0  ...    0   0    0   0   0   2   

   27    28    29    30  
0   0   0.0  33.0  33.0  
1  62   NaN   NaN   NaN  
2   0   5.0   8.0   0.0  
3  13  86.0   0.0   NaN  
4   0   0.0   0.0   0.0  

[5 rows x 31 columns]

In [17]:
df1=df.transpose()
df1.head()

0      1     2      3     4     5    6     7    8      9      10   11
0    0.0   61.0   2.0  260.0  47.0   0.0  0.0   0.0  0.0    0.0    1.0  0.0
1    0.0  103.0   0.0    1.0   0.0   0.0  0.0  45.0  0.0    0.0  163.0  0.0
2  272.0    0.0  17.0    0.0   0.0  30.0  0.0   0.0  0.0    0.0    5.0  0.0
3    0.0    2.0  27.0    0.0   0.0  42.0  0.0   0.0  0.0   14.0    0.0  0.0
4    0.0    0.0   0.0    0.0   5.0   0.0  0.0   0.0  0.0  156.0    0.0  0.0

By completing Example 3.3, you know the data in the excel file represents rainfall in _hundredths of an inch_, but for brevity's sake, this lab will treat the data as if it were rainfall in \textbf{inches}.  There is no need to change or scale the data before further processing.

Define a variable called ```month``` and set it equal to ```1``` for now.  This will represent the month for which we will compute statistics.  Later, you will change ```month``` to other values.  ```0``` means January, ```2``` means February, and so on.

Use the value in \textbf{month} to extract that month's data from your \textbf{wd}  matrix and store it in a new vector \textbf{m}.  Because January has 31 days, you should only see values in \textbf{m} that are positive or zero (no values of -99999 ).

For example, according to our data, the rainfall for the first 3 days in January are 0 inches, 0 inches, and 272 inches.  The average rain fall per day in the first 3 days of January is 90.667.  The average rain fall for the days of January 2nd and 3rd is 136.\\

Compute and store the following statistics/properties of the data in the dataframe ```df1``` in variables named as follows:

    \item \textbf{m\_avg}: average inches fallen in month \textbf{m}
    \item \textbf{m\_max}: maximum inches fallen in month \textbf{m}
    \item \textbf{d\_max}: day(s) of month \textbf{m} in which \textbf{m\_max} inches fell
    \item \textbf{d\_zero}: day(s) of month \textbf{m} in which zero inches fell
    \item \textbf{nd\_zero}: number of day(s) of month \textbf{m} in which zero inches fell
    \item \textbf{d\_nonzero}: day(s) of month \textbf{m} in which at least some rain fell
    \item \textbf{nd\_nonzero}: number of day(s) in month \textbf{m} in which at least some rain fell   
    \item \textbf{m\_avg\_nonzero}: average inches of only day(s) in which some rain fell
 

Test all of your values for the month of January manually to make sure they are correct.\\
\\
Now, improve your script to handle the non-day data points.  That is, set the \textbf{month} variable to 2 (February), which has multiple values equal to -99999, and make your script ignore any such values.  Note, ignoring the value -99999 this is not the same as setting -99999 to zero, as was done in Example 3.3.  Setting -99999 to zero invalidates any averages computed for such a month. Hint: assigning an empty matrix \textbf{[ ]} to a variable is different from assigning the variable the value zero.\\
\\
Run your improved script for the February data and check the results against manual calculations for that month.\\
\\
You may wish to test your script against other months' data.  When you are convinced your results and variable names are correct, go through and use the semicolon to suppress all screen output, then PUBLISH your .m-file as .html and upload your \textbf{lab2.html} file to D2L.\\
\\
For your own records, save your \textbf{lab2.m} file and your \textbf{weather\_data} file to the same directory.  You do not need to submit your \textbf{weather\_data} file because your instructor will have generated their own copy to test your script.\\
\\

## Deliverable

One .html-file (exported from a .ipynb-file) with output suppressed and except for the variables listed at the top of this page. The .m-file should run without errors. Uploaded the .html file to D2L by the end of the lab period.

#### _By D. Kruger, adapted by P. Kazarinoff, Portland Community College, 2018_